theano/gof/tests/test_vm.py

```diff
@@ -369,3 +369,93 @@ def check_storage(storage_map):
         assert check_storage(storage_map)[0]
         assert len(set([id(v) for v in
                         storage_map.values()])) < len(storage_map)
+
+
+def test_shape_reallocation_no_shape_infer():
+        a, b, c, d = [tensor.dvector(n) for n in ['a', 'b', 'c', 'd']]
+        z = (3*a + b)
+        x = (c + 4*d)
+        y = z + x
+
+        m = theano.compile.get_mode(theano.Mode(linker='vm_nogc'))
+        m = m.excluding('fusion', 'inplace')
+
+        f = theano.function([a, b, c, d], y, name="test_reduce_memory",
+                            mode=m)
+
+        output = f([1, 2], [3, 5], [9, 8], [4, 3])
+        assert output.any()
+        storage_map = f.fn.storage_map
+
+        def check_storage(storage_map):
+            from theano.tensor.var import TensorConstant
+            for i in storage_map.keys():
+                if not isinstance(i, TensorConstant):
+                    keys_copy = storage_map.keys()[:]
+                    keys_copy.remove(i)
+                    for o in keys_copy:
+                        if isinstance(storage_map[i][0], numpy.ndarray):
+                            storage_map_copy = storage_map[i][0].all()
+                        else:
+                            storage_map_copy = storage_map[i][0]
+                        if isinstance(storage_map_copy, numpy.ndarray):
+                            storage_map_copy = storage_map_copy.all()
+                        if isinstance(
+                            storage_map[i][0] == storage_map[o][0],
+                                numpy.ndarray):
+                            compare = (
+                                storage_map[i][0] == storage_map[o][0]).all()
+                        else:
+                            compare = (storage_map[i][0] == storage_map[o][0])
+                        if storage_map_copy and compare:
+                            return [True, storage_map[o][0]]
+            return [False, None]
+
+        assert check_storage(storage_map)[0]
+        assert len(set([id(v) for v in
+                        storage_map.values()])) < len(storage_map)
+
+
+def test_shape_reallocation_shape_infer():
```

Comment:
On this subject, the other test should test when the shape feature is disabled. For this, add
m=m.excluding('ShapeOpt')

This will change the result line 414. As both test have so much code in commun can you refactor them in 1 test to remove duplicate code?